Snapshots

MendelMonteiro opened the pull request:
This change consists of two new features.  The first is the ability to be notified of any new subscriptions and the second is to generate and send a snapshot to a peer when they create a new subscription.

ltrzesniewski commented on src/Abc.Zebus/Directory/PeerDirectoryClient.cs:
```diff
@@ -269,15 +275,62 @@ private void AddOrUpdatePeerEntry(PeerDescriptor peerDescriptor)
+            return _messageDispatcher.GetMessageHanlerInvokers()
+                                     .Select(x => x.MessageHandlerType.GetBaseTypes().SingleOrDefault(y => y.IsGenericType && y.GetGenericTypeDefinition() == typeof(SubscriptionHandler<>))?.GenericTypeArguments[0])
+                                     .Where(x => x != null)
+                                     .ToHashSet();
```
Subscriptions are updated *very* often, and this seems slow...

ltrzesniewski commented on src/Abc.Zebus/SubscriptionHandling/SubscriptionHandler.cs:
```diff
@@ -0,0 +1,27 @@
+    /// Extend this class with the message you wish to be notified of subscriptions for.
+    /// </summary>
+    /// <typeparam name="TMessage">The message that you wish to be notified for.</typeparam>
+    public abstract class SubscriptionHandler<TMessage> : IMessageHandler<SubscriptionUpdatedMessage>
```
Why not an interface instead?

MendelMonteiro replied:
Because we want to filter the messages that are handled

ltrzesniewski commented on src/Abc.Zebus/SubscriptionHandling/SubscriptionSnapshotGenerator.cs:
```diff
@@ -0,0 +1,38 @@
+            if (_bus is IInternalBus internalBus)
+                internalBus.Publish(snapshot, peerId);
+            else
+                throw new Exception("The bus is not an internal bus");
```
- Don't use `Exception` but a subtype
- *"internal bus"* does not mean anything to the user, the error message needs to be better

MendelMonteiro replied:
As above, the user will never see this message.

ltrzesniewski replied:
Same response as above 😉

ltrzesniewski commented on src/Abc.Zebus/SubscriptionHandling/SubscriptionHandler.cs:
```diff
@@ -0,0 +1,27 @@
+using Abc.Zebus.Directory;
+
+namespace Abc.Zebus.SubscriptionHandling
```
Maybe `Abc.Zebus.Subscriptions` would be a better namespace name? Or even use `Abc.Zebus.Directory`?

MendelMonteiro replied:
Agreed that `Subscriptions` is a better name for the namespace

ltrzesniewski commented on src/Abc.Zebus/SubscriptionHandling/SubscriptionUpdatedMessage.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Abc.Zebus.SubscriptionHandling
+{
+    public class SubscriptionUpdatedMessage : IMessage
```
`SubscriptionsUpdated` would be a better name (and in plural)

ltrzesniewski commented on src/Directory.Build.props:
```diff
@@ -12,8 +12,8 @@
   <PropertyGroup>
-    <ZebusVersion>3.5.0</ZebusVersion>
-    <ZebusContractsVersion>3.0.0</ZebusContractsVersion>
+    <ZebusVersion>3.5.1</ZebusVersion>
```
New feature = new minor version, but it would be better *not* to change the version number in a branch.

ltrzesniewski commented on src/Directory.Build.props:
```diff
@@ -12,8 +12,8 @@
-    <ZebusVersion>3.5.0</ZebusVersion>
-    <ZebusContractsVersion>3.0.0</ZebusContractsVersion>
+    <ZebusVersion>3.5.1</ZebusVersion>
+    <ZebusContractsVersion>3.0.1</ZebusContractsVersion>
```
`Zebus.Contracts` did *not* change (and it is almost never supposed to). Don't bump its version.

ltrzesniewski commented on src/Abc.Zebus.Testing/TestBus.cs:
```diff
@@ -79,15 +80,24 @@ public HashSet<Subscription> Subscriptions
+        {
+            Publish(message, (PeerId?) targetPeer);
+        }
+        private void Publish(IEvent message,  PeerId? targetPeer)
```
Empty lines between methods

ltrzesniewski commented on src/Abc.Zebus.Tests/Directory/PeerDirectoryClientTests.cs:
```diff
@@ -26,6 +30,58 @@ public partial class PeerDirectoryClientTests
+
+            public override void InvokeMessageHandler(IMessageHandlerInvocation invocation)
+            {
+                throw new NotImplementedException();
```
`NotImplementedException` shows as a TODO item in the IDE, please use something else (also for the other cases below)

ltrzesniewski commented on src/Abc.Zebus.Tests/SubscriptionHandling/SnapshotGeneratorTests.cs:
```diff
@@ -0,0 +1,53 @@
+        {
+        }
+
+        private class snapshotGeneratorTest : SubscriptionSnapshotGenerator<SnapshotTest, EventTest>
```
`SnapshotGeneratorTest` (upper case `S`)

MendelMonteiro replied:
🤦‍♂

ltrzesniewski commented on src/Abc.Zebus/SubscriptionHandling/SubscriptionHandler.cs:
```diff
@@ -0,0 +1,27 @@
+        /// </summary>
+        /// <param name="subscription">The newly created subscription</param>
+        /// <param name="peerId">The peer id of the peer that created the subscription</param>
+        protected abstract void OnSubscription(SubscriptionsForType subscription, PeerId peerId);
```
Should this be singular or plural?

MendelMonteiro replied:
Well spotted, it should be plural.

ltrzesniewski replied:
`OnSubscriptions` feels weird as a method name... `OnSubscriptionsUpdated` would be better IMO.

ltrzesniewski commented on src/Abc.Zebus/Directory/PeerDirectoryClient.cs:
```diff
@@ -27,13 +31,15 @@ public partial class PeerDirectoryClient : IPeerDirectory,
         private Peer _self;
 
-        public PeerDirectoryClient(IBusConfiguration configuration)
+        public PeerDirectoryClient(IBusConfiguration configuration, IMessageDispatcher messageDispatcher)
```
Not sure I like this dependency, it feels off. Wouldn't it be better if the bus forwarded this info from the peer directory through an event?

MendelMonteiro replied:
The `PeerDirectoryClient` also needs to know the subscribed message handlers in order to determine what should be raised through an event so I'm not sure how this could be implmeneted without injecting the `MessageDispatcher`.  I'm open to any suggestions though.

ltrzesniewski replied:
Maybe this should be the responsibility of the bus itself?

ltrzesniewski replied:
What do you think @ocoanet?

ltrzesniewski commented on src/Abc.Zebus/Directory/SubscriptionDefinition.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public class SubscriptionDefinition
+    {
+        public List<SubscriptionDefinitionPart> Parts { get; }
```
`IReadOnlyList<SubscriptionDefinitionPart>` or `SubscriptionDefinitionPart[]` would be better for a public API

Note to self: check later if this class needs to be public

MendelMonteiro replied:
How could it not be public it's returned as part of the public API on `Subscription`

ltrzesniewski replied:
That was just a note to self 😉

ltrzesniewski replied:
The comment on the property type is still valid though

ltrzesniewski commented on src/Abc.Zebus/Directory/SubscriptionsForType.cs:
```diff
@@ -38,6 +38,8 @@ public static SubscriptionsForType Create<TMessage>(params BindingKey[] bindingK
         public Subscription[] ToSubscriptions()
             => BindingKeys?.Select(bindingKey => new Subscription(MessageTypeId, bindingKey)).ToArray() ?? new Subscription[0];
 
+        public List<SubscriptionDefinition> GetDefinition() => BindingKeys.Select(x => new SubscriptionDefinition(MessageTypeId.GetMessageType(), x)).ToList();
```
`IReadOnlyList` or array would be better as this is a public API

MendelMonteiro commented on src/Abc.Zebus/Core/MessageContextAwareBus.cs:
```diff
@@ -91,6 +91,14 @@ public void Reply(IMessage response)
+            if (_bus is IInternalBus bus)
+                bus.Publish(message, targetPeer);
+            else
+                throw new NotImplementedException("Inner bus is not IInternalBus");
```
We should never provide a non-internal bus to the `MessageAwareContextBus` thus it's an error message designed for someone developing in the Zebus codebase.

ltrzesniewski replied:
It could still happen if the user overrides the bus in the container, but I agree it's not very likely.

The comment on the exception type is still valid.

ltrzesniewski commented on src/Abc.Zebus/Directory/PeerDirectoryClient.cs:
```diff
@@ -368,6 +434,8 @@ public void Handle(PeerSubscriptionsForTypesUpdated message)
 
             PeerUpdated?.Invoke(message.PeerId, PeerUpdateAction.Updated);
+
+            DispatchSubscriptionUpdatedMessages(message.PeerId, message.SubscriptionsForType); // TODO
```
What's this TODO?

MendelMonteiro replied:
It's recursive - one must delete the TODO

MendelMonteiro commented on src/Abc.Zebus/Directory/PeerDirectoryClient.cs:
```diff
@@ -294,56 +292,52 @@ public void Handle(PeerStarted message)
-                {
-                    if (!_snapshotGeneratingMessageTypes.Contains(subscription.MessageTypeId.GetMessageType()))
-                        continue;
+                if (!_snapshotGeneratingMessageTypes.Contains(subscription.MessageTypeId.GetMessageType()))
```
This will give incoherent results as the reference `_snapshotGeneratingMessageTypes` can change during the iteration.

ltrzesniewski commented:
We need to ensure the snapshots are transient.

ocoanet commented:
The feature introduced by this PR is a nice addition to Zebus. I like the overall design but I have a few comments.

1. This change introduces a dependency from the PeerDirectoryClient to the MessageDispatcher.
The PDC is also registering to an event raised by the MD, so there is now an extra bi-directional communication between the two classes. The Bus was already referencing the PDC and the MD. I think that it was more logical to use the Bus to coordinate the two classes. In addition, the Bus already know how to dispatch messages correctly and how to handle errors. With this PR, the PDC is now dispatching messages, but without proper error handling.

2. The SubscriptionHandler has no way to know what the new subscriptions are for a given message type. SubscriptionsUpdated contains all old and new subscriptions. Is it expected?

3. SubscriptionDefinition uses BindingKeyPredicateBuilder.GetRoutingMembers that is not cached.
I think it would be interesting to merge BindingKeyPredicateBuilder with MessageTypeDescriptor / MessageUtil. The MessageTypeDescriptor could include a collection of RoutingMember and other reflection based data.

4. SubscriptionDefinition is not really a subscription definition because it only contains the parts (the type is not exposed). In addition, SubscriptionDefinitionMatch introduces a new name / concept, but it can easily be merged with SubscriptionDefinitionPart.
I am reluctant to add three new types that have unclear usages and can be confusing in regards to the existing subscription types. Can we discuss this subject?

ltrzesniewski commented on src/Abc.Zebus/Core/Bus.cs:
```diff
@@ -769,6 +770,31 @@ private void HandleDeserializationError(MessageTypeId messageTypeId, Stream mess
+            }
+        }
+
+        private MessageContext GetMessageContextForSubscriptionsUpdated() => MessageContext.Current ?? MessageContext.CreateOverride(PeerId, EndPoint);
```
😢

ltrzesniewski commented on src/Abc.Zebus/Directory/PeerDirectoryClient.cs:
```diff
@@ -412,7 +367,18 @@ public void Handle(PeerSubscriptionsUpdated message)
+        {
+            var observedSubscriptions = peerDescriptorSubscriptions?
+                                        .Where(x => _observedSubscriptionMessageTypes.Contains(x.MessageTypeId.GetMessageType()))
+                                        .ToArray()
```
- `ToList` would be slightly better. We can still return `Array.Empty` when `peerDescriptorSubscriptions` is null.
- We should also directly return `Array.Empty` when `_observedSubscriptionMessageTypes` is empty (which will be true most of the time)

ltrzesniewski commented on src/Abc.Zebus/Directory/PeerDirectoryClient.cs:
```diff
@@ -431,7 +397,13 @@ public void Handle(PeerSubscriptionsForTypesUpdated message)
+                                               .Where(x => _observedSubscriptionMessageTypes.Contains(x.MessageTypeId.GetMessageType()))
+                                               .SelectMany(x => x.ToSubscriptions())
+                                               .ToArray()
+                                        ?? Array.Empty<Subscription>();
```
Same comment as above

ltrzesniewski commented on src/Abc.Zebus.Testing/Directory/TestPeerDirectory.cs:
```diff
@@ -46,7 +48,8 @@ public Task UnregisterAsync(IBus bus)
 
-        public void RegisterRemoteListener<TMEssage>() where TMEssage : IMessage
+        public void RegisterRemoteListener<TMEssage>()
+            where TMEssage : IMessage
```
Typo (not yours, but whatever)

ltrzesniewski commented:
> We need to ensure the snapshots are transient.

Update: We won't do this, because the `IMessageSendingStrategy` interface takes a `TransportMessage`, which we don't have in this context. Also, this feature is mainly intended for transient peers in the first place.

ocoanet commented:
@MendelMonteiro I just pushed a few changes to improve BindingKey API.

About SubscriptionDefinition:
- I removed SubscriptionDefinition / SubscriptionDefinitionPart.
- I added BindingKeyPart which is the new return type for BindingKey.GetPart.
- I added BindingKeyUtil.GetPartForMember.
- I added Subscription.GetBindingKeyPartForMember and SubscriptionsForType.GetBindingKeyPartsForMember.

Can you confirm that those new types and methods will be equivalent to help SubscriptionHandler and SubscriptionSnapshotGenerator implementers?

The other changes are mostly techniques, I moved all the reflection changes into MessageTypeDescriptor.

MendelMonteiro commented:
> @MendelMonteiro I just pushed a few changes to improve BindingKey API.
> 
> About SubscriptionDefinition:
> 
> * I removed SubscriptionDefinition / SubscriptionDefinitionPart.
> * I added BindingKeyPart which is the new return type for BindingKey.GetPart.
> * I added BindingKeyUtil.GetPartForMember.
> * I added Subscription.GetBindingKeyPartForMember and SubscriptionsForType.GetBindingKeyPartsForMember.
> 
> Can you confirm that those new types and methods will be equivalent to help SubscriptionHandler and SubscriptionSnapshotGenerator implementers?
> 
> The other changes are mostly techniques, I moved all the reflection changes into MessageTypeDescriptor.

This looks ok to me and the API is equivalent for the foreseen usage in the snapshot generators.
